### ISSUE TITLE
Hyper Signals / SMS / Email Alert

### ISSUE THREAD

**Issue by Flowery7** [Flowery7/Lottery-#20]
import requests
import pandas as pd
import time
import threading
from datetime import datetime, timedelta
import openai
from twilio.rest import Client
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

# Function to fetch price data
def fetch_price_data(api_url, api_key):
    headers = {'Authorization': f'Bearer {api_key}'}
    response = requests.get(api_url, headers=headers)
    data = response.json()
    return data

# Function to calculate SMA
def calculate_sma(prices, period):
    return prices.rolling(window=period).mean()

# Function to calculate EMA
def calculate_ema(prices, period):
    return prices.ewm(span=period, adjust=False).mean()

# Function to calculate RSI
def calculate_rsi(prices, period=14):
    delta = prices.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
    rs = gain / loss
    rsi = 100 - (100 / (1 + rs))
    return rsi

# Function to calculate MACD
def calculate_macd(prices, fast_period=12, slow_period=26, signal_period=9):
    ema_fast = calculate_ema(prices, fast_period)
    ema_slow = calculate_ema(prices, slow_period)
    macd = ema_fast - ema_slow
    signal = macd.ewm(span=signal_period, adjust=False).mean()
    histogram = macd - signal
    return macd, signal, histogram

# Function to determine buy/sell signals
def determine_signals(df):
    signals = []

    # SMA and EMA Crossover
    if df['SMA_20'].iloc[-1] > df['SMA_50'].iloc[-1] and df['SMA_20'].iloc[-2] <= df['SMA_50'].iloc[-2]:
        signals.append("Buy (SMA Crossover)")
    elif df['SMA_20'].iloc[-1] < df['SMA_50'].iloc[-1] and df['SMA_20'].iloc[-2] >= df['SMA_50'].iloc[-2]:
        signals.append("Sell (SMA Crossover)")

    if df['EMA_20'].iloc[-1] > df['EMA_50'].iloc[-1] and df['EMA_20'].iloc[-2] <= df['EMA_50'].iloc[-2]:
        signals.append("Buy (EMA Crossover)")
    elif df['EMA_20'].iloc[-1] < df['EMA_50'].iloc[-1] and df['EMA_20'].iloc[-2] >= df['EMA_50'].iloc[-2]:
        signals.append("Sell (EMA Crossover)")

    # RSI Levels
    if df['RSI_14'].iloc[-1] < 30:
        signals.append("Buy (RSI)")
    elif df['RSI_14'].iloc[-1] > 70:
        signals.append("Sell (RSI)")

    # MACD Crossover
    if df['MACD'].iloc[-1] > df['Signal'].iloc[-1] and df['MACD'].iloc[-2] <= df['Signal'].iloc[-2]:
        signals.append("Buy (MACD)")
    elif df['MACD'].iloc[-1] < df['Signal'].iloc[-1] and df['MACD'].iloc[-2] >= df['Signal'].iloc[-2]:
        signals.append("Sell (MACD)")

    return signals

# Function to fetch data and perform calculations
def data_fetch_and_calculate(api_url, api_key):
    global price_data
    while True:
        # Fetch current price data
        data = fetch_price_data(api_url, api_key)
        current_price = data['price']
        current_time = datetime.now()

        # Append current price data to list
        price_data.append({'timestamp': current_time, 'price': current_price})
        df = pd.DataFrame(price_data)

        # Calculate indicators if we have enough data
        if len(df) > 50:  # Ensure enough data points for calculations
            df['SMA_20'] = calculate_sma(df['price'], 20)
            df['SMA_50'] = calculate_sma(df['price'], 50)
            df['EMA_20'] = calculate_ema(df['price'], 20)
            df['EMA_50'] = calculate_ema(df['price'], 50)
            df['RSI_14'] = calculate_rsi(df['price'], 14)
            df['MACD'], df['Signal'], df['MACD_Hist'] = calculate_macd(df['price'])

            signals = determine_signals(df)
            if signals:
                message = f"Time: {current_time}, Price: {current_price}\n"
                for signal in signals:
                    message += f"Signal: {signal}\n"
                send_alert(message)

        # Wait for the next interval
        time.sleep(15 * 60)  # 15 minutes

# Function to fetch financial news and analyze with ChatGPT
def fetch_and_analyze_news(api_url, currency_pair, openai_api_key):
    global last_news_time
    while True:
        # Fetch relevant news using an internet search API (replace with actual API and method)
        params = {
            'q': currency_pair,
            'from': last_news_time.strftime('%Y-%m-%dT%H:%M:%SZ'),  # ISO 8601 format for datetime
            'sortBy': 'publishedAt',
            'language': 'en',
            'apiKey': openai_api_key  # Example usage, replace with actual news API key if needed
        }
        response = requests.get(api_url, params=params)
        news_data = response.json()
        
        # Filter for new articles
        news_articles = news_data.get('articles', [])
        relevant_news = [article['title'] + " " + article['description'] for article in news_articles if article['publishedAt'] > last_news_time.strftime('%Y-%m-%dT%H:%M:%SZ')]
        
        # Update the last news fetch time
        if news_articles:
            last_news_time = datetime.strptime(news_articles[0]['publishedAt'], '%Y-%m-%dT%H:%M:%SZ')

        # Analyze the news with ChatGPT if there is new relevant news
        if relevant_news:
            print("Analyzing news for potential impact...")
            for news in relevant_news:
                prompt = f"How might the following news affect the {currency_pair} currency pair? {news}"
                openai.api_key = openai_api_key
                response = openai.Completion.create(
                    engine="text-davinci-003",
                    prompt=prompt,
                    max_tokens=50
                )
                analysis = response.choices[0].text.strip()
                print(f"News: {
